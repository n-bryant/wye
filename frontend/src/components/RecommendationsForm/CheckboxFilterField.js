import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import Checkbox from "@material-ui/core/Checkbox";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import Typography from "@material-ui/core/Typography";

import styles from "./CheckboxFilterField.styles";

/**
 * renders a field for electing a selection of values in the form of a checkbox list
 */
export const CheckboxFilterField = props => {
  const classnames = CheckboxFilterField.classnames(props);
  const {
    classes,
    name,
    category,
    type,
    label,
    setFieldValue,
    value,
    options
  } = props;

  // set up initial state values
  const perPage = 20;
  let initialValuesState = {};
  options.forEach(option => {
    initialValuesState[option] = false;
  });
  const [checkedOptions, setCheckedOptions] = React.useState(
    initialValuesState
  );
  const [currentPage, setCurrentPage] = React.useState(1);

  /**
   * handle changes to the CheckBox options,
   * and update formik field value
   * @param {String} option
   */
  const handleChange = option => event => {
    const newValue = { ...checkedOptions, [option]: event.target.checked };
    setCheckedOptions(newValue);
    let valueToSet = {
      ...value,
      [category]: {
        ...value[category],
        [type]: Object.keys(newValue).filter(key => Boolean(newValue[key]))
      }
    };
    if (Object.keys(newValue).every(key => !newValue[key])) {
      delete valueToSet[category][type];
    }
    setFieldValue(name, valueToSet);
  };

  // handle scroll event and loading of more items
  const handleScroll = event => {
    const containerHeight = event.target.getBoundingClientRect().height;
    const optionsListHeight = event.target.firstElementChild.getBoundingClientRect()
      .height;
    if (event.target.scrollTop >= optionsListHeight - containerHeight - 84) {
      setCurrentPage(currentPage + 1);
    }
  };

  const optionsCount = Object.keys(checkedOptions).length;
  const selectedOptionsCount = Object.keys(checkedOptions).filter(
    key => !!checkedOptions[key]
  ).length;
  return (
    <div className={classnames.root()}>
      <Typography
        className={classnames.element("label")}
        variant="body2"
        gutterBottom={true}
      >
        {`${label} ${optionsCount} ${
          selectedOptionsCount > 0 ? `- (${selectedOptionsCount} selected)` : ""
        }`}
      </Typography>
      <div
        className={classnames.element("optionsContainer")}
        onScroll={handleScroll}
      >
        <div className={classnames.element("options")}>
          {options.slice(0, currentPage * perPage).map((option, index) => (
            <FormControlLabel
              key={index}
              control={
                <Checkbox
                  classes={{
                    root: classes.checkbox,
                    checked: classes.checkboxChecked
                  }}
                  checked={checkedOptions[option]}
                  color="default"
                  onChange={handleChange(option)}
                  value={option}
                  inputProps={{ "aria-label": `${name}-option_${option}` }}
                />
              }
              label={option}
            />
          ))}
        </div>
      </div>
    </div>
  );
};
CheckboxFilterField.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}CheckboxFilterField`
);
CheckboxFilterField.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    checkbox: PropTypes.string,
    label: PropTypes.string,
    optionsContainer: PropTypes.string,
    checkboxChecked: PropTypes.string
  }),
  // Formik field name
  name: PropTypes.string,
  // which filter category is the field for
  category: PropTypes.oneOf(["gameFilters", "playerFilters"]),
  // which filter type is it
  type: PropTypes.oneOf([
    "ownedBy",
    "recentlyPlayedBy",
    "publishers_in",
    "developers_in",
    "genres_in",
    "tags_in"
  ]).isRequired,
  // the label to display for the field
  label: PropTypes.string,
  // playerFilters field value from Formik form
  value: PropTypes.object,
  // handler for updating users field value
  setFieldValue: PropTypes.func,
  // the available steamids to select
  options: PropTypes.array
};
CheckboxFilterField.defaultProps = {
  classes: {}
};

// apply styles
export default withStyles(styles)(CheckboxFilterField);
