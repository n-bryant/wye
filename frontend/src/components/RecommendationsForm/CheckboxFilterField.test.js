import FormControlLabel from "@material-ui/core/FormControlLabel";
import { CheckboxFilterField } from "./CheckboxFilterField";

describe("CheckboxFilterField", () => {
  const props = {
    classes: {
      root: "root",
      checkbox: "checkbox",
      label: "label",
      optionsContainer: "optionsContainer",
      checkboxChecked: "checkboxChecked"
    },
    name: "name",
    category: "gameFilters",
    label: "label",
    type: "ownedBy",
    value: {
      gameFilters: {}
    },
    setFieldValue: jest.fn(),
    options: ["optionA", "optionB"]
  };

  let wrapper;
  beforeEach(() => {
    const checkedOptions = {
      optionA: false,
      optionB: false
    };
    const setCheckedOptions = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [checkedOptions, setCheckedOptions]);
    wrapper = shallow(<CheckboxFilterField {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should update the formik field value when a change occurs", () => {
    shallow(
      wrapper
        .find(FormControlLabel)
        .at(0)
        .prop("control")
    ).prop("onChange")({ target: { checked: true } });
    expect(props.setFieldValue).toHaveBeenCalledWith(props.name, {
      gameFilters: {
        [props.type]: ["optionA"]
      }
    });
  });
});
