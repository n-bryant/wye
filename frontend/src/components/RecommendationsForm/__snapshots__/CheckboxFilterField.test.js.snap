// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxFilterField should render successfully 1`] = `
<div
  className="wye-CheckboxFilterField root"
>
  <WithStyles(ForwardRef(Typography))
    className="wye-CheckboxFilterField__label label"
    gutterBottom={true}
    variant="body2"
  >
    label 2 
  </WithStyles(ForwardRef(Typography))>
  <div
    className="wye-CheckboxFilterField__optionsContainer optionsContainer"
  >
    <WithStyles(ForwardRef(FormControlLabel))
      control={
        <ForwardRef(WithStyles)
          checked={false}
          classes={
            Object {
              "checked": "checkboxChecked",
              "root": "checkbox",
            }
          }
          color="default"
          inputProps={
            Object {
              "aria-label": "name-option_optionA",
            }
          }
          onChange={[Function]}
          value="optionA"
        />
      }
      key="0"
      label="optionA"
    />
    <WithStyles(ForwardRef(FormControlLabel))
      control={
        <ForwardRef(WithStyles)
          checked={false}
          classes={
            Object {
              "checked": "checkboxChecked",
              "root": "checkbox",
            }
          }
          color="default"
          inputProps={
            Object {
              "aria-label": "name-option_optionB",
            }
          }
          onChange={[Function]}
          value="optionB"
        />
      }
      key="1"
      label="optionB"
    />
  </div>
</div>
`;
